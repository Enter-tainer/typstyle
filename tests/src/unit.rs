use std::{env, error::Error, fs, path::Path};

use insta::internals::Content;
use libtest_mimic::{Failed, Trial};
use typst_syntax::Source;
use typstyle_core::Typstyle;

use crate::common::{fixtures_dir, read_source_with_options};

/// Creates one test for each `.typ` file in the current directory or
/// sub-directories of the current directory.
pub fn collect_tests() -> Result<Vec<Trial>, Box<dyn Error>> {
    fn make_snapshot_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_snapshot(&path, width)
        })
        .with_kind("snapshot")
    }

    fn make_convergence_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_convergence(&path, width)
        })
        .with_kind("convergence")
    }

    #[cfg(feature = "consistency")]
    fn make_consistency_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_output_consistency(&path, width)
        })
        .with_kind("consistency")
    }

    fn visit_dir(path: &Path, tests: &mut Vec<Trial>) -> Result<(), Box<dyn Error>> {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let file_type = entry.file_type()?;
            let path = entry.path();

            if file_type.is_dir() {
                // Handle directories
                if path.file_name() == Some("partial".as_ref()) {
                    // Ignore partial tests
                    continue;
                }
                visit_dir(&path, tests)?;
                continue;
            } else if !(file_type.is_file() && path.extension() == Some("typ".as_ref())) {
                continue;
            }
            // Handle .typ files
            let name = path
                .strip_prefix(env::current_dir()?)?
                .display()
                .to_string();

            let is_no_snap = path
                .file_stem()
                .and_then(|name| name.to_str())
                .is_some_and(|name| name.starts_with('_'));
            if !is_no_snap {
                tests.extend([
                    make_snapshot_test(&path, &name, 0),
                    make_snapshot_test(&path, &name, 40),
                    make_snapshot_test(&path, &name, 80),
                    make_snapshot_test(&path, &name, 120),
                ]);
            }
            tests.extend([
                make_convergence_test(&path, &name, 0),
                make_convergence_test(&path, &name, 40),
                make_convergence_test(&path, &name, 80),
            ]);
            #[cfg(feature = "consistency")]
            tests.extend([
                make_consistency_test(&path, &name, 0),
                make_consistency_test(&path, &name, 40),
                make_consistency_test(&path, &name, 80),
            ]);
        }

        Ok(())
    }

    // We recursively look for `.typ` files, starting from the current directory.
    let mut tests = Vec::new();
    let current_dir = fixtures_dir();
    visit_dir(&current_dir, &mut tests)?;

    Ok(tests)
}

fn check_snapshot(path: &Path, width: usize) -> Result<(), Failed> {
    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;

    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);
    settings.set_omit_expression(true);
    settings.set_snapshot_path(path.parent().unwrap().join("snap"));
    settings.set_input_file(path);
    if source.root().erroneous() {
        settings.set_raw_info(&Content::Map(vec![("erroneous".into(), true.into())]));
    }
    settings.bind(|| {
        let snap_name = format!("{}-{width}", path.file_name().unwrap().to_str().unwrap());
        if source.root().erroneous() {
            insta::assert_snapshot!(snap_name, "");
        } else {
            cfg.max_width = width;
            let mut formatted = Typstyle::new(cfg).format_source(&source).unwrap();
            if formatted.starts_with('\n') {
                formatted.insert_str(0, "// DUMMY\n");
            }
            if formatted.ends_with("\n\n") {
                formatted.push_str("// DUMMY\n");
            }

            insta::assert_snapshot!(snap_name, formatted);
        }
    });
    Ok(())
}

fn check_convergence(path: &Path, width: usize) -> Result<(), Failed> {
    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;
    if source.root().erroneous() {
        return Ok(());
    }

    cfg.max_width = width;
    let mut first_pass = Typstyle::new(cfg.clone()).format_source(&source)?;
    for i in 0..=opt.relax_convergence {
        let new_source = Source::detached(&first_pass);
        if new_source.root().erroneous() {
            panic!(
                "the source becomes erroneous after {} iterations:\n{:#?}",
                i + 1,
                new_source.root().errors()
            )
        }
        let second_pass = Typstyle::new(cfg.clone()).format_source(&new_source)?;
        if first_pass == second_pass {
            return Ok(());
        }
        if i == opt.relax_convergence {
            pretty_assertions::assert_str_eq!(
                first_pass,
                second_pass,
                "formatting does not converge in {} iterations!",
                opt.relax_convergence
            );
        }
        first_pass = second_pass;
    }

    Ok(())
}

#[cfg(feature = "consistency")]
fn check_output_consistency(path: &Path, width: usize) -> Result<(), Failed> {
    use std::path::PathBuf;

    use typstyle_consistency::{ErrorSink, FormattedSources, FormatterHarness};

    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;
    if source.root().erroneous() {
        return Ok(());
    }

    cfg.max_width = width;

    let mut err_sink = ErrorSink::new(format!("consistency {}", path.display()));

    let mut harness = FormatterHarness::new("".to_string(), PathBuf::new())?;
    let main_vpath = Path::new("__main__");
    harness.add_source_file(main_vpath, source.text())?;

    let base_world = harness.snapshot();
    let fmt_sources = FormattedSources {
        name: "formatted".to_string(),
        sources: harness.format(
            &base_world,
            |source| Ok(Typstyle::new(cfg.clone()).format_source(&source).unwrap()),
            &mut err_sink,
        )?,
    };

    harness.compile_and_compare([fmt_sources].iter(), main_vpath, false, &mut err_sink)?;

    if err_sink.is_ok() {
        Ok(())
    } else {
        Err(err_sink.into())
    }
}
