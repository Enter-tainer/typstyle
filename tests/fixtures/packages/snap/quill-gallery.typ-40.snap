---
source: tests/src/unit.rs
input_file: tests/fixtures/packages/quill-gallery.typ
---
/// typstyle: wrap_text
/*
 * This test case is copied from [quill]
 * Original source: [https://github.com/Mc-Zen/quill/blob/main/docs/guide/gallery.typ]
 * Author: [Mc-Zen]
 * License: [MIT]
 *
 * This file is used as test case only and only the import is changed from the original.
 */

#import "@preview/quill:0.7.0": *

#set page(
  width: 18cm,
  height: auto,
  margin: 0mm,
)


#let gallery = {
  set raw(lang: "typc")
  table(
    align: center + horizon,
    columns: (
      1fr,
      1fr,
      1.3fr,
      1.1fr,
      1fr,
      1.48fr,
    ),
    column-gutter: (
      0pt,
      0pt,
      2.5pt,
      0pt,
      0pt,
    ),
    [Normal gate],
    quantum-circuit(1, gate($H$), 1),
    raw("gate($H$), $H$"),
    [Round gate],
    quantum-circuit(
      1,
      gate($X$, radius: 100%),
      1,
    ),
    raw("gate($X$, \nradius: 100%)"),

    [D-shaped gate],
    quantum-circuit(
      1,
      gate($Y$, radius: (right: 100%)),
      1,
    ),
    raw(
      "gate($Y$, radius: \n(right: 100%))",
    ),
    [Meter],
    quantum-circuit(1, meter(), 1),
    raw("meter()"),

    [Meter with \ label],
    quantum-circuit(
      1,
      meter(label: $lr(|±〉)$),
      1,
    ),
    raw("meter(label: \n$lr(|±〉)$)"),
    [Phase gate],
    quantum-circuit(1, phase($α$), 1),
    raw("phase($α$)"),

    [Control],
    quantum-circuit(1, ctrl(), 1),
    raw("ctrl()"),
    [Open control],
    quantum-circuit(
      1,
      ctrl(open: true),
      1,
    ),
    raw("ctrl(open: true)"),

    [Target],
    quantum-circuit(1, targ(), 1),
    raw("targ()"),
    [Swap target],
    quantum-circuit(1, swap(), 1),
    raw("swap()"),

    [Permutation \ gate],
    quantum-circuit(
      1,
      permute(2, 0, 1),
      1,
      [\ ],
      3,
      [\ ],
      3,
    ),
    raw("permute(2,0,1)"),
    [Multi-qubit \ gate],
    quantum-circuit(
      1,
      mqgate($U$, n: 3),
      1,
      [\ ],
      3,
      [\ ],
      3,
    ),
    raw("mqgate($U$, n: 3)"),

    [lstick],
    quantum-circuit(
      lstick($|psi〉$),
      2,
    ),
    raw("lstick($|psi〉$)"),
    [rstick],
    quantum-circuit(2, rstick(
      $|psi〉$,
    )),
    raw("rstick($|psi〉$)"),

    [Multi-qubit \ lstick],
    quantum-circuit(
      row-spacing: 10pt,
      lstick($|psi〉$, n: 2),
      2,
      [\ ],
      3,
    ),
    raw("lstick($|psi〉$, \nn: 2)"),
    [Multi-qubit \ rstick],
    quantum-circuit(
      row-spacing: 10pt,
      2,
      rstick(
        $|psi〉$,
        n: 2,
        brace: "]",
      ),
      [\ ],
      3,
    ),
    raw(
      "rstick($|psi〉$, \nn: 2, brace: \"]\")",
    ),

    [midstick],
    quantum-circuit(
      1,
      midstick("yeah"),
      1,
    ),
    raw("midstick(\"yeah\")"),
    [Wire bundle],
    quantum-circuit(1, nwire(5), 1),
    raw("nwire(5)"),

    [Controlled \ #smallcaps("z")-gate],
    quantum-circuit(
      1,
      ctrl(1),
      1,
      [\ ],
      1,
      ctrl(),
      1,
    ),
    [#raw("ctrl(1)") \ + \ #raw(
        "ctrl()",
      )],
    [Controlled \ #smallcaps("x")-gate],
    quantum-circuit(
      1,
      ctrl(1),
      1,
      [\ ],
      1,
      targ(),
      1,
    ),
    [#raw("ctrl(1)") \ + \ #raw(
        "targ()",
      )],

    [Swap \ gate],
    quantum-circuit(
      1,
      swap(1),
      1,
      [\ ],
      1,
      swap(),
      1,
    ),
    [#raw("swap(1)") \ + \ #raw(
        "swap()",
      )],
    [Controlled \ Hadamard],
    quantum-circuit(
      1,
      mqgate($H$, target: 1),
      1,
      [\ ],
      1,
      ctrl(),
      1,
    ),
    [#raw("mqgate($H$,target:1)") \ + \
      #raw("ctrl()")],

    [Plain\ vertical\ wire],
    quantum-circuit(
      1,
      ctrl(1, show-dot: false),
      1,
      [\ ],
      3,
    ),
    raw("ctrl(1, show-dot: false)"),
    [Meter to \ classical],
    quantum-circuit(
      1,
      meter(target: 1),
      1,
      [\ ],
      setwire(2),
      1,
      ctrl(),
      1,
    ),
    [#raw("meter(target: 1)") \ + \
      #raw("ctrl()")],

    [Classical wire],
    quantum-circuit(setwire(2), 3),
    raw("setwire(2)"),
    [Styled wire],
    quantum-circuit(
      setwire(1, stroke: green),
      3,
    ),
    raw("setwire(1, stroke: green)"),

    [Labels],
    quantum-circuit(
      scale: 100%,
      1,
      gate($Q$, label: (
        (content: "b", pos: top),
        (content: "b", pos: bottom),
        (content: "a", pos: left + top),
        (
          content: "c",
          pos: right + top,
          dy: 0pt,
          dx: 50%,
        ),
      )),
      1,
    ),
    [#set text(size: .6em);```typc
      gate($Q$, label: (
      (content: "b",pos:top),
      (content:"b",pos:bottom),
      ( content: "a",
        pos: left + top ),
      ( content: "c",
        pos: right + top,
        dy: 0pt, dx: 50% ),
      ))
      ```
    ],
    [Gate inputs \ and outputs],
    quantum-circuit(
      scale: 80%,
      1,
      mqgate(
        $U$,
        n: 3,
        width: 5em,
        inputs: (
          (qubit: 0, n: 2, label: $x$),
          (qubit: 2, label: $y$),
        ),
        outputs: (
          (qubit: 0, n: 2, label: $x$),
          (qubit: 2, label: $y ⊕ f(x)$),
        ),
      ),
      1,
      [\ ],
      3,
      [\ ],
      3,
    ),
    [#set text(size: .6em);```typc
      mqgate($U$, n: 3, width: 5em,
        inputs: (
          (qubit:0, n:2, label:$x$),
          (qubit:2, label: $y$)
        ),
        outputs: (
          (qubit:0, n:2, label:$x$),
          (qubit:2, label:$y⊕f(x)$)
        )
      )```],
  )
}


#rect(
  stroke: none,
  radius: 3pt,
  inset: (x: 6pt, y: 6pt),
  fill: white,
  gallery,
)
